rust/client-shared/src/eventloop.rs

```diff
@@ -569,3 +602,130 @@ fn is_unreachable(e: &io::Error) -> bool {
         || e.kind() == io::ErrorKind::HostUnreachable
         || e.kind() == io::ErrorKind::AddrNotAvailable
 }
+
+struct UdpDnsClient {
+    socket_factory: Arc<dyn SocketFactory<UdpSocket>>,
+    servers: Vec<SocketAddr>,
+}
+
+impl UdpDnsClient {
+    const TIMEOUT: Duration = Duration::from_secs(2);
+
+    fn new(socket_factory: Arc<dyn SocketFactory<UdpSocket>>) -> Self {
+        Self {
+            socket_factory,
+            servers: Vec::default(),
+        }
+    }
+
+    async fn resolve(&self, host: String) -> Result<Vec<IpAddr>> {
+        let host = DomainName::vec_from_str(&host).context("Failed to parse domain name")?;
+        let servers = self.servers.clone();
+
+        let (a_records, aaaa_records) = self
+            .servers
+            .iter()
+            .map(|socket| {
+                futures::future::try_join(
+                    self.send(
+                        *socket,
+                        dns_types::Query::new(host.clone(), dns_types::RecordType::A),
+                    ),
+                    self.send(
+                        *socket,
+                        dns_types::Query::new(host.clone(), dns_types::RecordType::AAAA),
+                    ),
+                )
+                .map_err(|e| {
+                    tracing::debug!(%host, "DNS query failed: {e:#}");
+
+                    e
+                })
+            })
+            .collect::<FuturesUnordered<_>>()
+            .filter_map(|result| result.ok())
+            .filter(|(a, b)| {
+                a.response_code() == dns_types::ResponseCode::NOERROR
+                    && b.response_code() == dns_types::ResponseCode::NOERROR
+            })
+            .next()
+            .await
+            .with_context(|| {
+                format!("All DNS servers ({servers:?}) failed to resolve portal host '{host}'")
+            })?;
+
+        let ips = iter::empty()
```

Comment:
do we need more than one IP resolved here? 

if we care about the first one, perhaps something like https://docs.rs/futures/0.3.31/futures/future/fn.select_ok.html could do ? instead of waiting for all of them to resolve?

Reply:
Yeah we definitely want to use all IPs that we resolve. `select_ok` across all DNS servers could be an interesting choice though. We'd have to see how we can incorporate the checking of the response code to avoid having those used.